cpp/src/arrow/compute/kernels/scalar_hash.cc

```diff
@@ -0,0 +1,194 @@
+namespace compute {
+namespace internal {
+
+// Define symbols visible within `arrow::compute::internal` in this file;
+// these symbols are not visible outside of this file.
+namespace {
+
+// ------------------------------
+// Function documentation
+const FunctionDoc fast_hash_32_doc{
+    "Construct a hash for every element of the input argument",
+    ("An element-wise function that uses an xxHash-like algorithm.\n"
+     "This function is not suitable for cryptographic purposes.\n"
+     "Hash results are 32-bit and emitted for each valid row.\n"
+     "Null (or invalid) rows emit a null in the output."),
+    {"hash_input"}};
+
+const FunctionDoc fast_hash_64_doc{
+    "Construct a hash for every element of the input argument",
+    ("An element-wise function that uses an xxHash-like algorithm.\n"
+     "This function is not suitable for cryptographic purposes.\n"
+     "Hash results are 64-bit and emitted for each valid row.\n"
+     "Null (or invalid) rows emit a null in the output."),
+    {"hash_input"}};
+
+// ------------------------------
+// Kernel implementations
+struct FastHash32Scalar {
+  static Status Exec(KernelContext* ctx, const ExecSpan& input_arg, ExecResult* out) {
+    if (input_arg.num_values() != 1 or not input_arg[0].is_array()) {
+      return Status::Invalid("FastHash32 currently supports a single array input");
+    }
+
+    // Initialize stack-based memory allocator with an allocator and memory size
+    util::TempVectorStack stack_memallocator;
+    ARROW_RETURN_NOT_OK(stack_memallocator.Init(ctx->exec_context()->memory_pool(),
+                                                max_bitwidth * max_batchsize));
+
+    // Prepare input data structure for propagation to hash function
+    ArraySpan hash_input = input_arg[0].array;
+    ARROW_ASSIGN_OR_RAISE(KeyColumnArray input_keycol,
+                          ColumnArrayFromArrayData(hash_input.ToArrayData(),
+                                                   default_rstart, hash_input.length));
+
+    // Call hashing function
+    std::vector<uint32_t> hash_results(hash_input.length);
+    LightContext hash_ctx;
+
+    hash_ctx.hardware_flags = ctx->exec_context()->cpu_info()->hardware_flags();
+    hash_ctx.stack = &stack_memallocator;
+    Hashing32::HashMultiColumn({input_keycol}, &hash_ctx, hash_results.data());
+
+    // Prepare results of hash function for kernel output argument
+    arrow::UInt32Builder builder;
+    ARROW_RETURN_NOT_OK(builder.Reserve(hash_results.size()));
+    ARROW_RETURN_NOT_OK(builder.AppendValues(hash_results));
+    ARROW_ASSIGN_OR_RAISE(auto result_array, builder.Finish());
+
+    out->value = ArraySpan{*(result_array->data())};
+    return Status::OK();
+  }
+
+  // This 96 represents the most bits *per row* that the Hashing32 and Hashing64
+  // algorithms allocate from the provided `TempVectorStack`
+  static constexpr uint32_t max_bitwidth = 96;
+  static constexpr uint32_t max_batchsize = util::MiniBatch::kMiniBatchLength;
```

Comment:
Same with this constant, I'd just use `util::MiniBatch::kMiniBatchLength` directly in the code.